dm/pkg/checker/table_structure.go

```diff
@@ -299,19 +343,96 @@ func (c *TablesChecker) checkUnique(cst *ast.Constraint) bool {
 }
 
-func (c *TablesChecker) checkTableOption(opt *ast.TableOption) *incompatibilityOption {
-	if opt.Tp == ast.TableOptionCharset {
-		// Check charset
-		cs := strings.ToLower(opt.StrValue)
-		if cs != "binary" && !charset.ValidCharsetAndCollation(cs, "") {
-			return &incompatibilityOption{
-				state:       StateWarning,
-				instruction: "https://docs.pingcap.com/tidb/stable/mysql-compatibility#unsupported-features",
-				errMessage:  fmt.Sprintf("unsupport charset %s", opt.StrValue),
+func (c *TablesChecker) checkTableStructurePair(upstream, downstream *ast.CreateTableStmt) []*incompatibilityOption {
+	//nolint: prealloc
+	var options []*incompatibilityOption
+
+	// check charset of upstream/downstream tables
+	upstreamCharset := getCharset(upstream)
+	downstreamCharset := getCharset(downstream)
+	if upstreamCharset != "" && downstreamCharset != "" &&
+		!strings.EqualFold(upstreamCharset, downstreamCharset) &&
+		!strings.EqualFold(downstreamCharset, mysql.UTF8MB4Charset) {
+		options = append(options, &incompatibilityOption{
+			state: StateWarning,
+			errMessage: fmt.Sprintf("charset is not same, upstream: (%s %s), downstream: (%s %s)",
+				upstream.Table.Name.O, upstreamCharset,
+				downstream.Table.Name.O, downstreamCharset),
+		})
+	}
+
+	// check collation
+	upstreamCollation := getCollation(upstream)
+	downstreamCollation := getCollation(downstream)
+	if upstreamCollation != "" && downstreamCollation != "" &&
+		!strings.EqualFold(upstreamCollation, downstreamCollation) {
+		options = append(options, &incompatibilityOption{
+			state: StateWarning,
+			errMessage: fmt.Sprintf("collation is not same, upstream: (%s %s), downstream: (%s %s)",
+				upstream.Table.Name.O, upstreamCollation,
+				downstream.Table.Name.O, downstreamCollation),
+		})
+	}
+
+	// check PK/UK
+	upstreamPKUK := getPKAndUK(upstream)
+	downstreamPKUK := getPKAndUK(downstream)
+	// the number of PK/UK should be small, we use a simple but slow algorithm for now
+	for idxNameUp, s := range upstreamPKUK {
+		for idxNameDown, s2 := range downstreamPKUK {
+			if stringSetEqual(s, s2) {
+				delete(upstreamPKUK, idxNameUp)
+				delete(downstreamPKUK, idxNameDown)
+				break
 			}
 		}
 	}
-	return nil
+	for idxName, cols := range upstreamPKUK {
+		options = append(options, &incompatibilityOption{
+			state: StateWarning,
+			errMessage: fmt.Sprintf("upstream has more PK or NOT NULL UK than downstream, index name: %s, columns: %v",
+				idxName, utils.SetToSlice(cols)),
+		})
+	}
+	for idxName, cols := range downstreamPKUK {
+		options = append(options, &incompatibilityOption{
+			state: StateWarning,
+			errMessage: fmt.Sprintf("downstream has more PK or NOT NULL UK than upstream, table name: %s, index name: %s, columns: %v",
+				downstream.Table.Name.O, idxName, utils.SetToSlice(cols)),
+		})
+	}
+
+	// check columns
+	upstreamCols := getColumnsAndIgnorable(upstream)
+	downstreamCols := getColumnsAndIgnorable(downstream)
+	// TODO: handle extended column
```

Comment:
Maybe we can't be released in 6.4 until this is fixed?

Reply:
The user should be able to use that feature according to the document, we just can't find the error in advance because precheck is not implemented. I think it's OK.

Reply:
Extended columns feature downstream will always have more columns. Will this check always return a warning?

Reply:
Currently yes, I'll fix this TODO before v6.5.0 release